crypto/rsa/rsa_backend.c

```diff
@@ -64,22 +64,109 @@ static int collect_numbers(STACK_OF(BIGNUM) *numbers,
+    int derive_from_pq = 0;
+    BN_CTX *ctx = NULL;
 
     if (rsa == NULL)
         return 0;
 
+    ctx = BN_CTX_new_ex(rsa->libctx);
+    if (ctx == NULL)
+        return 0;
+
+    param_derive = OSSL_PARAM_locate_const(params,
+                                           OSSL_PKEY_PARAM_RSA_DERIVE_FROM_PQ);
+    if ((param_derive != NULL)
+        && !OSSL_PARAM_get_int(param_derive, &derive_from_pq))
+        goto err;
+
     param_n = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_N);
     param_e = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_E);
     if (include_private)
         param_d = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_D);
 
-    if ((param_n != NULL && !OSSL_PARAM_get_BN(param_n, &n))
-        || (param_e != NULL && !OSSL_PARAM_get_BN(param_e, &e))
-        || (param_d != NULL && !OSSL_PARAM_get_BN(param_d, &d)))
-        goto err;
+    if (derive_from_pq) {
+        /* we need at minimum p, q and e */
+        param_p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_FACTOR1);
+        param_q = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_FACTOR2);
+        if ((param_p != NULL && !OSSL_PARAM_get_BN(param_p, &p))
+            || (param_q != NULL && !OSSL_PARAM_get_BN(param_q, &q))
+            || (param_e != NULL && !OSSL_PARAM_get_BN(param_e, &e))) {
+            ERR_raise(ERR_LIB_RSA, ERR_R_PASSED_NULL_PARAMETER);
+            goto err;
+        }
+
+        /* compute n if its not provided */
+        if (param_n == NULL) {
+            n = BN_new();
+            if (n == NULL) {
+                ERR_raise(ERR_LIB_RSA, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
+            if (!BN_mul(n, p, q, ctx)) {
+                ERR_raise(ERR_LIB_RSA, ERR_R_INTERNAL_ERROR);
+                goto err;
+            }
+        } else {
+            if (!OSSL_PARAM_get_BN(param_n, &n)) {
+                ERR_raise(ERR_LIB_RSA, ERR_R_PASSED_INVALID_ARGUMENT);
+                goto err;
+            }
+        }
+
+        /* compute d if it is not provided, and private key requested */
+        if (include_private && param_d == NULL) {
+            BIGNUM *lcm, *p1q1, *gcd, *p1, *q1;
+
+            lcm = BN_CTX_get(ctx);
+            p1q1 = BN_CTX_get(ctx);
+            gcd = BN_CTX_get(ctx);
+            p1 = BN_CTX_get(ctx);
+            q1 = BN_CTX_get(ctx);
+
+            if (lcm == NULL || p1q1 == NULL || gcd == NULL
+                || p1 == NULL || q1 == NULL) {
+                ERR_raise(ERR_LIB_RSA, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
+
+            d = BN_new();
+            if (d == NULL) {
+                ERR_raise(ERR_LIB_RSA, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
+
+            if (ossl_rsa_get_lcm(ctx, p, q, lcm, gcd, p1, q1, p1q1) != 1) {
+                ERR_raise(ERR_LIB_RSA, ERR_R_INTERNAL_ERROR);
+                goto err;
+            }
+
+            BN_set_flags(d, BN_FLG_CONSTTIME);
```

Comment:
Should p and q have this set too??

Reply:
I think we're ok without the extra cals to BN_set_flags, as prior to any use of p or q, we assign them to the RSA key with RSA_set0_factors, which sets the flag for us

Reply:
Hm... you seem to call RSA_set0_factors only later. Perhaps all of this code computing n should be moved after the CRT param derivation?

Reply:
I wouldn't avoid calling `BN_set_flags()` even if it seems excessive because it could be a side effect of another function: that function could change and the results could be catastrophic.

This is the cost of having `BN_FLG_CONSTTIME` as opt-in rather than default.